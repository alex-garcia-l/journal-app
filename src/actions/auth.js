import Swal from 'sweetalert2';

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { firebase, googleAuthProvider } from '../database/firebase-config';

export const startLogin = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLoginWithGoogle = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  }
}

export const startUserRegister = (name, email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})

export const logout = () => ({
  type: types.logout
})
