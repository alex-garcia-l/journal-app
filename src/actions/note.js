import Swal from 'sweetalert2';
import { db } from '../database/firebase-config';
import { types } from '../types/types';
import { loadNotes } from '../helpers/loadNotes';
import { uploadImage } from '../helpers/uploadImage';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(setNoteActive(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  }
}

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const startUpdateNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteUpdate = { ...note };
    delete noteUpdate.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteUpdate);

    dispatch(refreshNotes(note.id, noteUpdate));

    Swal.fire({
      title: 'Success',
      text: 'Note modified',
      icon: 'success',
      showConfirmButton: false,
      timer: 1500
    });
  }
}

export const startUploadImage = (file) => {
  return async (dispatch, getState) => {
    const { active: note } = getState().notes;

    Swal.fire({
      title: 'Uploading',
      text: 'Please wait...',
      allowOutsideClick: false,
      allowEscapeKey: false,
      allowEnterKey: false,
      didOpen: () => {
        Swal.showLoading();
      },
      willClose: () => {
        Swal.hideLoading()
      },
    });

    const upload = await uploadImage(file);
    note.url = upload.secure_url;
    dispatch(startUpdateNote(note));

    Swal.close();
  }
}

export const startDeleteNote = (id) => {
  return (dispatch, getState) => {

    Swal.fire({
      title: 'Are you sure you want to delete the note?',
      confirmButtonText: 'Yes, delete',
      showCancelButton: true,
      icon: 'question',
    }).then(async (result) => {
      /* Read more about isConfirmed, isDenied below */
      if (result.isConfirmed) {
        const { uid } = getState().auth;

        await db.doc(`${uid}/journal/notes/${id}`).delete();
        dispatch(deleteNote(id));
      }
    })
  }
}

export const setNoteActive = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
});

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
});

export const refreshNotes = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note
    }
  }
});

export const purgeNotes = () => ({
  type: types.notesLogoutCleaning
})
