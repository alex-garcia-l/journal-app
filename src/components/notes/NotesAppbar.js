import React from 'react';

export const NotesAppbar = () => {
  return (
    <div className="notes__appbar">
      <span>23 sep 2020</span>
      <div>
        <button
          className="btn btn-primary mr-5"
        >
          Picture
        </button>

        <button
          className="btn btn-primary"
        >
          Save
        </button>
      </div>
    </div>
  );
};
