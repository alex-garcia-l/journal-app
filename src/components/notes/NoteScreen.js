import { useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { setNoteActive, startDeleteNote } from '../../actions/note';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {

  const dispatch = useDispatch();
  const { active: note } = useSelector(state => state.notes);
  let [formValues, handleInputChange, resetForm] = useForm(note);
  const { title, body, id } = formValues;
  const activeId = useRef(note.id)

  useEffect(() => {
    if (note.id !== activeId.current) {
      resetForm(note);
      activeId.current = note.id
    }
  }, [note, resetForm]);

  useEffect(() => {
    dispatch(setNoteActive(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  const handleDelete = () => {
    dispatch(startDeleteNote(id));
  }

  return (
    <div className="notes__main-content animate__animated animate__fadeIn">
      <NotesAppBar />

      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          autoComplete="off"
          name="title"
          onChange={handleInputChange}
          value={title}
        />

        <textarea
          placeholder="What happened today"
          className="notes__title-textarea"
          autoComplete="off"
          name="body"
          onChange={handleInputChange}
          value={body}
        ></textarea>

        {
          note.url &&
          (
            <div className="notes__image">
              <img
                src={note.url}
                alt={title}
              />
            </div>
          )
        }
      </div>
      <button
        className="btn btn-third"
        onClick={handleDelete}
      >
        Delete
      </button>
    </div>
  );
};
