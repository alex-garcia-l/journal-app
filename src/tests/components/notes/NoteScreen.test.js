/**
 * @jest-environment node
 */

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { setNoteActive } from '../../../actions/note';

import 'jsdom-global/register';

jest.mock('../../../actions/note', () => ({
  setNoteActive: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const id = 'bkqgO47fuY5fLRHqrMTO';

const initialState = {
  auth: {
    id: '123',
    name: 'Nombre de Usuario'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    active: {
      id,
      title: 'Title',
      body: 'Body',
      date: 1648183494210
    },
    notes: [{
      id: "bkqgO47fuY5fLRHqrMTO",
      title: "",
      body: "",
      date: 1648183494210
    }],
  }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <NoteScreen />
  </Provider >
);

describe('Pruebas en <NoteScreen />', () => {

  test('debe mostrarse correctamente', () => {

    expect(wrapper).toMatchSnapshot();

  });

  test('debe de llamar setNoteActive correctamente', () => {

    const newTitle = 'Body Test';

    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: newTitle
      }
    });

    expect(setNoteActive).toHaveBeenLastCalledWith(id, {
      id,
      title: newTitle,
      body: 'Body',
      date: 1648183494210
    })

  });

});
