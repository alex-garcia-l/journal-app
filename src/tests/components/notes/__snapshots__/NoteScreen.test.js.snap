// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn"
>
  <div
    className="notes__appbar"
  >
    <span>
      March 24th 2022, 10:44:54 pm
    </span>
    <input
      id="uploadFile"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn btn-primary mr-5"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn btn-primary"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Title"
    />
    <textarea
      autoComplete="off"
      className="notes__title-textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Body"
    />
  </div>
  <button
    className="btn btn-third"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
